icicle/backend/cpu/src/hash/cpu_keccak.cpp

```diff
@@ -1,119 +1,319 @@
+ * SHA3-256, SHA3-384, SHA-512 are implemented. SHA-224 can easily be added.
+ * Based on code from http://keccak.noekeon.org/ .
+ * Original work can be found on https://github.com/brainhub/SHA3IUF
+ * ---------------------------------------------------------------------- */
 
 #include "icicle/backend/hash/keccak_backend.h"
+#include "icicle/utils/modifiers.h"
 
 namespace icicle {
 
-  class KeccakBackend : public HashBackend
+// Configuration flags for Keccak and SHA-3
+#define SHA3_USE_KECCAK_FLAG (0x80000000)
+#define SHA3_CW(x)           ((x) & (~SHA3_USE_KECCAK_FLAG))
+#ifndef SHA3_ROTL64
+#define SHA3_ROTL64(x, y) (((x) << (y)) | ((x) >> ((sizeof(uint64_t) * 8) - (y))))
+#endif
+
+// Define sponge words for Keccak
+#define SHA3_KECCAK_SPONGE_WORDS (((1600) / 8) / sizeof(uint64_t))
+
+  class KeccakBackendCPU : public HashBackend
   {
   public:
-    KeccakBackend(uint64_t input_chunk_size, uint64_t output_size, uint64_t rate, int padding_const, const char* name)
-        : HashBackend(name, output_size, input_chunk_size)
+    KeccakBackendCPU(uint64_t input_chunk_size, uint64_t output_size, bool is_keccak, const char* name)
+        : HashBackend{name, output_size, input_chunk_size}, m_is_keccak{is_keccak}
     {
     }
 
     eIcicleError hash(const std::byte* input, uint64_t size, const HashConfig& config, std::byte* output) const override
     {
-      ICICLE_LOG_DEBUG << "Keccak/sha3 CPU hash() called, batch=" << config.batch
-                       << ", single_output_size=" << output_size();
-      // TODO implement real logic
-      for (int i = 0; i < output_size() * config.batch; ++i) {
-        output[i] = std::byte(i % 256);
+      const auto digest_size_in_bytes = output_size();
+      const auto single_input_size = get_single_chunk_size(
+        size); // if size==0 using default input chunk size. This is useful for Merkle-Tree constructions
+      ICICLE_LOG_DEBUG << name() << "::hash() called, batch=" << config.batch << ", input-size=" << size << " bytes";
+
+      // TODO (future): use tasks manager to parallel across threads. Add option to config-extension to set #threads
+      // with default=0. for now we don't do it and let the merkle-tree define the parallelizm so hashing a large batch
+      // outside a merkle-tree context is not as fast as it could be.
+      // Note that for batch=1 this has not effect.
+      for (unsigned batch_idx = 0; batch_idx < config.batch; ++batch_idx) {
+        eIcicleError err = sha3_hash_buffer(
+          8 * digest_size_in_bytes /*=bitsize*/, m_is_keccak, input + batch_idx * single_input_size, single_input_size,
+          output + batch_idx * digest_size_in_bytes);
+
+        if (err != eIcicleError::SUCCESS) { return err; }
       }
       return eIcicleError::SUCCESS;
     }
+
+  private:
+    static const uint64_t s_keccakf_rndc[24];
+    static const unsigned s_keccakf_rotc[24];
+    static const unsigned s_keccakf_piln[24];
+    const bool m_is_keccak;
```

Comment:
what is that memeber for?

Reply:
The implementation that Aviad used needs to know that since it's both for keccak and sha3.
Since the same object implements keccak and sha3, it is stored. maybe an enum is better.